subprojects/logging/src/main/java/org/gradle/internal/featurelifecycle/DeprecationWarningBuildOperation.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.featurelifecycle;
+
+import org.gradle.internal.operations.BuildOperationContext;
+import org.gradle.internal.operations.BuildOperationDescriptor;
+import org.gradle.internal.operations.RunnableBuildOperation;
+
+import java.util.List;
+
+import static org.gradle.internal.featurelifecycle.DeprecationWarningBuildOperationType.*;
+
+public class DeprecationWarningBuildOperation implements RunnableBuildOperation {
+    private static final Object RESULT = new Result() {};
+
+    private FeatureUsage usage;
+
+    public DeprecationWarningBuildOperation(FeatureUsage usage) {
+        this.usage = usage;
+    }
+
+    @Override
+    public BuildOperationDescriptor.Builder description() {
+        return BuildOperationDescriptor.displayName("Deprecation Warning - " + usage.getMessage())
+            .details(new Details() {
+                public String getMessage() {
+                    return usage.getMessage();
+                }
+
+                @Override
+                public List<StackTraceElement> getStackTrace() {
+                    return usage.withStackTrace().getStack();
```

Comment:
This is going to lead to traces to the call of this method, and effectively makes this type mutable. Check the implementation of withStackTrace().

Reply:
+1